jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/WebSocketDemander.java

```diff
@@ -97,21 +141,43 @@ public void setNextDemand(DemandChain nextDemand)
      */
     public void onFrame(Frame frame, Callback callback)
     {
-        if (_frame != null || _callback != null)
-            throw new IllegalStateException("Not expecting onFrame");
+        Throwable failure = null;
+        try (AutoLock l = _lock.lock())
+        {
+            if (_failure != null)
+            {
+                failure = _failure;
+            }
+            else if (_inputDemand)
+            {
+                // Fulfill the incoming demand to receive a new frame.
+                _inputDemand = false;
+                _frame = frame;
+                _callback = new CountingCallback(callback, 1);
```

Comment:
No need to wrap with `CountingCallback` is the count is `1`, no?

I see the comment in `onCompleteFailure()`, but either that comment should be here, or better we should try to avoid to double-complete the callback (this should be controlled, as this is an `IteratingCallback` so it should never be double-completed).

Reply:
Currently if the `WebSocketDemander` is failed, we fail the callback from the `_callback` field, but the application might also have that callback which it can complete.

We cannot control this through the IteratingCallback because the application does not need to succeed the callback to get the next frame, they just call demand.